File: npc_sim/npc/npc_factory.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Fluent factory for creating pre-configured NPC archetypes."""

from __future__ import annotations
from npc_sim.core.sim_rng import SimRng
from npc_sim.npc.npc import NPC
from npc_sim.npc.identity import NPCIdentity
from npc_sim.npc.vitals import NPCVitals
from npc_sim.npc.psychology import NPCPsychology
from npc_sim.npc.social import NPCSocial
from npc_sim.npc.traits import NPCTraits


class NPCFactory:
    """Fluent factory for creating pre-configured NPCs. Fully deterministic via SimRng."""

    @staticmethod
    def create(
        display_name: str, age: int, rng: SimRng,
        gender: str = "Unknown", occupation: str = "Civilian",
        faction: str = "None", personality_archetype: str = "Generic",
        max_health: float = 100.0, max_energy: float = 100.0,
        memory_capacity: int = 50,
    ) -> NPC:
        identity = NPCIdentity(
            rng.next_id("npc"), display_name, age,
            gender, occupation, faction, personality_archetype)
        return NPC(identity, NPCVitals(max_health, max_energy),
                   NPCPsychology(), NPCSocial(), memory_capacity)

    @staticmethod
    def create_guard(rng: SimRng, name: str = None, faction: str = "City Watch") -> NPC:
        name = name or "Guard"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(22, 45),
            gender="Male", occupation="Guard", faction=faction,
            personality_archetype="Guardian")
        vitals = NPCVitals(max_health=120.0, max_energy=110.0)
        psych = NPCPsychology(
            extraversion=rng.next_float(0.3, 0.5),
            agreeableness=rng.next_float(0.3, 0.6),
            conscientiousness=rng.next_float(0.7, 1.0),
            neuroticism=rng.next_float(0.1, 0.4),
            openness=rng.next_float(0.1, 0.4))
        traits = NPCTraits(NPCTraits.BRAVE, NPCTraits.CAUTIOUS)
        return NPC(identity, vitals, psych, NPCSocial(), memory_capacity=60, traits=traits)

    @staticmethod
    def create_merchant(rng: SimRng, name: str = None, faction: str = "Merchant Guild") -> NPC:
        name = name or "Merchant"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(25, 60),
            gender="Unknown", occupation="Merchant", faction=faction,
            personality_archetype="Merchant")
        psych = NPCPsychology(
            extraversion=rng.next_float(0.6, 0.9),
            agreeableness=rng.next_float(0.5, 0.8),
            conscientiousness=rng.next_float(0.4, 0.7),
            neuroticism=rng.next_float(0.2, 0.5),
            openness=rng.next_float(0.5, 0.8))
        traits = NPCTraits(NPCTraits.GREEDY, NPCTraits.CURIOUS)
        return NPC(identity, NPCVitals(), psych, NPCSocial(), memory_capacity=70, traits=traits)

    @staticmethod
    def create_civilian(rng: SimRng, name: str = None, faction: str = "None") -> NPC:
        name = name or "Civilian"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(18, 70),
            gender="Unknown", occupation="Civilian", faction=faction,
            personality_archetype="Generic")
        psych = NPCPsychology(
            extraversion=rng.next_float(0.3, 0.7),
            agreeableness=rng.next_float(0.4, 0.8),
            conscientiousness=rng.next_float(0.3, 0.7),
            neuroticism=rng.next_float(0.3, 0.7),
            openness=rng.next_float(0.3, 0.6))
        traits = NPCTraits(NPCTraits.CAUTIOUS)
        return NPC(identity, NPCVitals(), psych, NPCSocial(), traits=traits)

    @staticmethod
    def create_scholar(rng: SimRng, name: str = None, faction: str = "Academy") -> NPC:
        name = name or "Scholar"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(28, 65),
            gender="Unknown", occupation="Scholar", faction=faction,
            personality_archetype="Scholar")
        psych = NPCPsychology(
            extraversion=rng.next_float(0.2, 0.5),
            agreeableness=rng.next_float(0.5, 0.8),
            conscientiousness=rng.next_float(0.7, 1.0),
            neuroticism=rng.next_float(0.2, 0.5),
            openness=rng.next_float(0.8, 1.0))
        vitals = NPCVitals(max_health=80.0, max_energy=90.0)
        traits = NPCTraits(NPCTraits.CURIOUS, NPCTraits.PACIFIST)
        return NPC(identity, vitals, psych, NPCSocial(), memory_capacity=100, traits=traits)

    @staticmethod
    def create_farmer(rng: SimRng, name: str = None, faction: str = "Farmers Guild") -> NPC:
        name = name or "Farmer"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(18, 60),
            gender="Unknown", occupation="Farmer", faction=faction,
            personality_archetype="Farmer")
        psych = NPCPsychology(
            extraversion=rng.next_float(0.3, 0.6),
            agreeableness=rng.next_float(0.5, 0.8),
            conscientiousness=rng.next_float(0.5, 0.8),
            neuroticism=rng.next_float(0.2, 0.5),
            openness=rng.next_float(0.4, 0.7))
        traits = NPCTraits(NPCTraits.CAUTIOUS)
        return NPC(identity, NPCVitals(90.0, 120.0), psych, NPCSocial(), memory_capacity=50, traits=traits)

    @staticmethod
    def create_priest(rng: SimRng, name: str = None, faction: str = "Temple") -> NPC:
        name = name or "Priest"
        identity = NPCIdentity(
            rng.next_id("npc"), name, age=rng.next_int(30, 70),
            gender="Unknown", occupation="Priest", faction=faction,
            personality_archetype="Priest")
        psych = NPCPsychology(
            extraversion=rng.next_float(0.4, 0.7),
            agreeableness=rng.next_float(0.7, 1.0),
            conscientiousness=rng.next_float(0.6, 0.9),
            neuroticism=rng.next_float(0.1, 0.4),
            openness=rng.next_float(0.6, 1.0))
        traits = NPCTraits("Devout", NPCTraits.PACIFIST)
        return NPC(identity, NPCVitals(85.0, 90.0), psych, NPCSocial(), memory_capacity=70, traits=traits)

File: Runtime/NPC/NPCFactory.cs
using System;

namespace ForgeProject.Sim
{
    /// <summary>
    /// Fluent factory for creating pre-configured NPCs.
    /// Uses <see cref="SimRng"/> for all ID and stat generation — fully deterministic.
    /// </summary>
    public static class NPCFactory
    {
        // ─── Generic Builder ──────────────────────────────────────────────────────

        /// <summary>
        /// Creates a minimal NPC with the given name and age.
        /// All other stats use defaults.
        /// </summary>
        public static NPC Create(
            string   displayName,
            int      age,
            SimRng   rng,
            string   gender              = "Unknown",
            string   occupation          = "Civilian",
            string   faction             = "None",
            string   personalityArchetype = "Generic",
            float    maxHealth           = 100f,
            float    maxEnergy           = 100f,
            int      memoryCapacity      = 50)
        {
            var identity = new NPCIdentity(
                rng.NextId("npc"), displayName, age,
                gender, occupation, faction, personalityArchetype);

            return new NPC(
                identity,
                new NPCVitals(maxHealth, maxEnergy),
                new NPCPsychology(),
                new NPCSocial(),
                memoryCapacity);
        }

        // ─── Presets ──────────────────────────────────────────────────────────────

        /// <summary>
        /// A resilient, alert guard. High conscientiousness and low openness.
        /// Traits: Brave, Cautious.
        /// </summary>
        public static NPC CreateGuard(SimRng rng, string name = null, string faction = "City Watch")
        {
            name ??= "Guard";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(22, 45),
                "Male", "Guard", faction, "Guardian");

            var vitals = new NPCVitals(maxHealth: 120f, maxEnergy: 110f);

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.3f, 0.5f),
                agreeableness:     rng.NextFloat(0.3f, 0.6f),
                conscientiousness: rng.NextFloat(0.7f, 1.0f),
                neuroticism:       rng.NextFloat(0.1f, 0.4f),
                openness:          rng.NextFloat(0.1f, 0.4f));

            var traits = new NPCTraits(NPCTraits.Brave, NPCTraits.Cautious);
            return new NPC(identity, vitals, psych, new NPCSocial(), memoryCapacity: 60, traits: traits);
        }

        /// <summary>
        /// An opportunistic merchant. High extraversion and agreeableness.
        /// Traits: Greedy, Curious.
        /// </summary>
        public static NPC CreateMerchant(SimRng rng, string name = null, string faction = "Merchant Guild")
        {
            name ??= "Merchant";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(25, 60),
                "Unknown", "Merchant", faction, "Merchant");

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.6f, 0.9f),
                agreeableness:     rng.NextFloat(0.5f, 0.8f),
                conscientiousness: rng.NextFloat(0.4f, 0.7f),
                neuroticism:       rng.NextFloat(0.2f, 0.5f),
                openness:          rng.NextFloat(0.5f, 0.8f));

            var traits = new NPCTraits(NPCTraits.Greedy, NPCTraits.Curious);
            return new NPC(identity, new NPCVitals(), psych, new NPCSocial(), memoryCapacity: 70, traits: traits);
        }

        /// <summary>
        /// A cautious civilian. Balanced stats with higher neuroticism.
        /// Traits: Cautious.
        /// </summary>
        public static NPC CreateCivilian(SimRng rng, string name = null, string faction = "None")
        {
            name ??= "Civilian";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(18, 70),
                "Unknown", "Civilian", faction, "Generic");

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.3f, 0.7f),
                agreeableness:     rng.NextFloat(0.4f, 0.8f),
                conscientiousness: rng.NextFloat(0.3f, 0.7f),
                neuroticism:       rng.NextFloat(0.3f, 0.7f),
                openness:          rng.NextFloat(0.3f, 0.6f));

            var traits = new NPCTraits(NPCTraits.Cautious);
            return new NPC(identity, new NPCVitals(), psych, new NPCSocial(), traits: traits);
        }

        /// <summary>
        /// A scholar: high openness and conscientiousness, lower physical stats.
        /// Traits: Curious, Pacifist.
        /// </summary>
        public static NPC CreateScholar(SimRng rng, string name = null, string faction = "Academy")
        {
            name ??= "Scholar";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(28, 65),
                "Unknown", "Scholar", faction, "Scholar");

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.2f, 0.5f),
                agreeableness:     rng.NextFloat(0.5f, 0.8f),
                conscientiousness: rng.NextFloat(0.7f, 1.0f),
                neuroticism:       rng.NextFloat(0.2f, 0.5f),
                openness:          rng.NextFloat(0.8f, 1.0f));

            var vitals = new NPCVitals(maxHealth: 80f, maxEnergy: 90f);
            var traits = new NPCTraits(NPCTraits.Curious, NPCTraits.Pacifist);
            return new NPC(identity, vitals, psych, new NPCSocial(), memoryCapacity: 100, traits: traits);
        }

        /// <summary>
        /// A hard-working farmer. Early dawn schedule, high conscientiousness.
        /// Traits: Cautious.
        /// </summary>
        public static NPC CreateFarmer(SimRng rng, string name = null, string faction = "Farmers Guild")
        {
            name ??= "Farmer";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(18, 60),
                "Unknown", "Farmer", faction, "Farmer");

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.3f, 0.6f),
                agreeableness:     rng.NextFloat(0.5f, 0.8f),
                conscientiousness: rng.NextFloat(0.5f, 0.8f),
                neuroticism:       rng.NextFloat(0.2f, 0.5f),
                openness:          rng.NextFloat(0.4f, 0.7f));

            var traits = new NPCTraits(NPCTraits.Cautious);
            return new NPC(identity, new NPCVitals(90f, 120f), psych, new NPCSocial(), memoryCapacity: 50, traits: traits);
        }

        /// <summary>
        /// A devout priest. High agreeableness and openness, spiritual focus.
        /// Traits: Devout, Pacifist.
        /// </summary>
        public static NPC CreatePriest(SimRng rng, string name = null, string faction = "Temple")
        {
            name ??= "Priest";
            var identity = new NPCIdentity(
                rng.NextId("npc"), name, age: rng.NextInt(30, 70),
                "Unknown", "Priest", faction, "Priest");

            var psych = new NPCPsychology(
                extraversion:      rng.NextFloat(0.4f, 0.7f),
                agreeableness:     rng.NextFloat(0.7f, 1.0f),
                conscientiousness: rng.NextFloat(0.6f, 0.9f),
                neuroticism:       rng.NextFloat(0.1f, 0.4f),
                openness:          rng.NextFloat(0.6f, 1.0f));

            var traits = new NPCTraits("Devout", NPCTraits.Pacifist);
            return new NPC(identity, new NPCVitals(85f, 90f), psych, new NPCSocial(), memoryCapacity: 70, traits: traits);
        }
    }
}
